bson/unmarshal_value_test.go

```diff
@@ -0,0 +1,108 @@
+		for _, tc := range unmarshalValueTestCases {
+			t.Run(tc.name, func(t *testing.T) {
+				gotValue := reflect.New(reflect.TypeOf(tc.val))
+				err := UnmarshalValue(tc.bsontype, tc.bytes, gotValue.Interface())
+				assert.Nil(t, err, "UnmarshalValueWithRegistry error: %v", err)
+				assert.Equal(t, tc.val, gotValue.Elem().Interface(), "value mismatch; expected %s, got %s", tc.val, gotValue.Elem())
+			})
+		}
+	})
+	t.Run("UnmarshalValueWithRegistry with DefaultRegistry", func(t *testing.T) {
+		for _, tc := range unmarshalValueTestCases {
+			t.Run(tc.name, func(t *testing.T) {
+				gotValue := reflect.New(reflect.TypeOf(tc.val))
+				err := UnmarshalValueWithRegistry(DefaultRegistry, tc.bsontype, tc.bytes, gotValue.Interface())
+				assert.Nil(t, err, "UnmarshalValueWithRegistry error: %v", err)
+				assert.Equal(t, tc.val, gotValue.Elem().Interface(), "value mismatch; expected %s, got %s", tc.val, gotValue.Elem())
+			})
+		}
+	})
+	// tests covering GODRIVER-2779
+	t.Run("UnmarshalValueWithRegistry with custom registry", func(t *testing.T) {
+		testCases := []struct {
+			name     string
+			val      interface{}
+			bsontype bsontype.Type
+			bytes    []byte
+		}{
+			{
+				name:     "SliceCodec binary",
+				val:      []byte("hello world"),
+				bsontype: bsontype.Binary,
+				bytes:    bsoncore.AppendBinary(nil, bsontype.BinaryGeneric, []byte("hello world")),
+			},
+			{
+				name:     "SliceCodec string",
+				val:      []byte("hello world"),
+				bsontype: bsontype.String,
+				bytes:    bsoncore.AppendString(nil, "hello world"),
+			},
+		}
+		rb := NewRegistryBuilder().RegisterTypeDecoder(reflect.TypeOf([]byte{}), bsoncodec.NewSliceCodec()).Build()
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				gotValue := reflect.New(reflect.TypeOf(tc.val))
+				err := UnmarshalValueWithRegistry(rb, tc.bsontype, tc.bytes, gotValue.Interface())
+				assert.Nil(t, err, "UnmarshalValueWithRegistry error: %v", err)
+				assert.Equal(t, tc.val, gotValue.Elem().Interface(), "value mismatch; expected %s, got %s", tc.val, gotValue.Elem())
+			})
+		}
+	})
+}
+
+// tests covering GODRIVER-2779
+func BenchmarkSliceCodecUnmarshal(b *testing.B) {
+	benchmarks := []struct {
+		name     string
+		bsontype bsontype.Type
+		bytes    []byte
+	}{
+		{
+			name:     "SliceCodec binary",
+			bsontype: bsontype.Binary,
+			bytes:    bsoncore.AppendBinary(nil, bsontype.BinaryGeneric, []byte(strings.Repeat("t", 4096))),
+		},
+		{
+			name:     "SliceCodec string",
+			bsontype: bsontype.String,
+			bytes:    bsoncore.AppendString(nil, strings.Repeat("t", 4096)),
+		},
+	}
+	rb := NewRegistryBuilder().RegisterTypeDecoder(reflect.TypeOf([]byte{}), bsoncodec.NewSliceCodec()).Build()
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.RunParallel(func(pb *testing.PB) {
+				for pb.Next() {
+					var gotValue []byte
+					err := UnmarshalValueWithRegistry(rb, bm.bsontype, bm.bytes, &gotValue)
```

Comment:
Optional, define assignment object inline since we don't use gotValue.

```suggestion
					err := UnmarshalValueWithRegistry(rb, bm.bsontype, bm.bytes, &[]byte{})
```